### DISCUSSION TITLE
Making dataset smaller

### DISCUSSION THREAD

**Discussion by Johan154** [talmolab/sleap#1044] [Help!] [answered]
Dear all,

For my project, I labeled about 1000 frames. However, I want to test the model's performance when it uses less training data and when it uses only a specific part of the training data. However, when you delete labeled frames in a project, you delete them from the labeling suggestions, but they are still in the project. What is the solution to completely deleting them? Or should I take another approach?

**Comment by roomrys** [comment, marked as answer]
Hi @Johan154,

As you pointed out, deleting a frame from the Labeling Suggestions just removes the frame from the list of suggestions and does not actually delete the labeled frame from the project. 

Side note
---

We have been working on implementing a feature to add a "Labels Table" #736 to SLEAP (similar to the Labeling Suggestions, but displays all labeled frames in the project and allows removal of labeled frames). The current consensus is that we need to restructure how we store labeled frames in order to be able to quickly display and sort frames for large projects.
> See the branches [liezl/add-labels-table](https://github.com/talmolab/sleap/tree/liezl/add-labels-table) and [liezl/labels-table-cache](https://github.com/talmolab/sleap/tree/liezl/labels-table-cache) for the progress.

TLDR
---

Without the Labels Table, you will need to either 
1. use the GUI to delete all instances in the frame (Labels > Custom Instance Delete) or 
2. use the SLEAP API, see instructions below

## Deleting Select Labeled Frames using the SLEAP API
First activate the `sleap` conda environment, ensure `ipython` is installed, and run `ipython`
```
conda activate sleap
pip install ipython
ipython
```
Once we are in ipython, we can execute the following commands
```python
import sleap

ds = "path/to/slp"  # CHANGE THIS to be the path to your slp file
labels = sleap.load_file(ds)  # All the data for the slp project is stored here

labels.videos  # This prints all the videos in the slp project (use this to set video_idx below)
video_idx = 0  # CHANGE THIS to be the index of the video to delete labeled frames from
video = labels.videos[video_idx]

# Note that frame_idx starts at 0 in the API, but starts at 1 in the GUI
labeled_frame_indices = [0, 3, 255, 344]  # CHANGE THIS to be the list of frame indices to delete
for frame_idx in labeled_frame_indices:
    # Note that frame_idx starts at 0 in the API, but starts at 1 in the GUI
    lf_to_remove = labels.get(video, frame_idx)  # Get the labeled frame in video at frame_idx
    labels.remove(lf_to_remove)  # Remove the labeled frame from the project

new_slp = "path/to/modified/slp"  # Use a different name from the original slp - to be safe
labels.save(new_slp)  # Save the modified slp
```

---

All the answers in this thread are great!

**Reply by amblypatty** [reply]
In my case, I couldn't extract easily the labeled frame indices from the SLEAP GUI to put into the `labeled_frames_indices` list. I wanted to delete all the labeled frames from a specific video in my dataset. If you don't know the exact frame numbers that you are planning to delete, you could get a list of frame indices using the following list comprehension command
```
labeled_frame_indices = [lf.frame_idx for lf in labels.labeled_frames if lf.video.filename == 'path/to/video'] # list of frame indices to delete from a video in dataset
```
This creates a list of frame numbers from a specific video. Just delete the conditional `if` statement if you want to get frame numbers from the entire project.

**Reply by talmo** [reply]
Hi friends,

Another suggestion: if you want to remove all frames associated with a video, you can use [`Labels.remove_video`](https://sleap.ai/api/sleap.io.dataset.html#sleap.io.dataset.Labels.remove_video) to do something like this:

```python
import sleap

labels = sleap.load_file("labels.slp")

video_to_delete = labels.videos[4]  # Or whatever you want to remove.

labels.remove_video(video_to_delete)  # Removes associated labels as well.

labels.save("new_labels.slp")
```
